nomt/src/beatree/ops/update/mod.rs

```diff
@@ -50,231 +50,344 @@ pub fn update(
-            .map(|(k, _)| k);
-
-        // first leaf cutoff is the separator of the second leaf _or_ the separator of the next
-        // branch if there is only 1 leaf, or nothing.
-        let first_leaf_cutoff = first_branch
-            .as_ref()
-            .and_then(|node| {
-                if node.n() > 1 {
-                    Some(reconstruct_key(node.prefix(), node.separator(1)))
-                } else {
-                    None
-                }
-            })
-            .or(first_branch_cutoff);
-
-        let first_leaf = first_branch
-            .as_ref()
-            .map(|node| {
-                (
-                    PageNumber::from(node.node_pointer(0)),
-                    reconstruct_key(node.prefix(), node.separator(0)),
-                )
-            })
-            .map(|(id, separator)| BaseLeaf {
-                id,
-                node: leaf_pages.remove(&id).unwrap_or_else(|| LeafNode {
-                    inner: ctx.leaf_reader.query(id),
-                }),
-                iter_pos: 0,
-                separator,
-            });
+impl LeafChanges {
+    fn delete(&mut self, key: Key, pn: PageNumber) {
+        let entry = self.inner.entry(key).or_insert_with(|| ChangedLeafEntry {
+            deleted: None,
+            inserted: None,
+        });
 
-        // start with the first branch, cut-off second branch key.
-        let branch_updater = BranchUpdater::new(
-            first_branch.map(|node| BaseBranch {
-                // UNWRAP: node can only exist if ID does.
-                id: first.as_ref().unwrap().1,
-                node,
-                iter_pos: 0,
-            }),
-            first_branch_cutoff,
-        );
-
-        // start with first leaf in first branch, cut-off second leaf key.
-        let leaf_updater = LeafUpdater::new(first_leaf, first_leaf_cutoff);
-
-        Updater {
-            branch_updater,
-            leaf_updater,
-            obsolete_branches: Vec::new(),
-            leaf_pages,
-        }
+        // we can only delete a leaf once.
+        assert!(entry.deleted.is_none());
+
+        entry.deleted = Some(pn);
     }
 
-    fn ingest(&mut self, key: Key, value_change: Option<Vec<u8>>, ctx: &mut Ctx) {
-        self.digest_until(Some(key), ctx);
+    fn insert(&mut self, key: Key, node: LeafNode) {
+        let entry = self.inner.entry(key).or_insert_with(|| ChangedLeafEntry {
+            deleted: None,
+            inserted: None,
+        });
 
-        let (value_change, overflow) = if let Some(ref large_value) = value_change
-            .as_ref()
-            .filter(|v| v.len() > MAX_LEAF_VALUE_SIZE)
-        {
-            let pages = overflow::chunk(large_value, &mut ctx.leaf_writer);
-            (Some(overflow::encode_cell(large_value.len(), &pages)), true)
-        } else {
-            (value_change, false)
-        };
+        if let Some(_prev) = entry.inserted.replace(node) {
+            // TODO: this is where we'd clean up.
```

Comment:
this TODO isn't for this PR per se, but rather for when we'd have a leaf-page cache.